README.md
# miot
Modular Internet of Things

STATUS:  PRE-ALPHA  ** UNSTABLE DO NOT INSTALL **

VERSION: 0.2.1
    - WARNING - Unstable development version
    - Not feature complete
    - Under development
    - Not ready for internal testing

A Pre-Alpha release:
* Is still under development and considered unstable
* May have had limited or no internal testing.
* May not be feature complete
* May have features that are partially complete or not working. 
* May have known or unknown bugs
* May operate in an unexpected way. 
* May contain output debugging data to logs or screen.
* Is NOT supported: Use at your own risk.

Please do not raise tickets for bugs found in this release.

# INFORMATION

This is the command line control shell for all modules.

# FEATURES:

* No manual editing of config files.
* Configure all MIoT modules from one simple shell
* install/uninstall other modules
* start/stop and get status of installed modules

# PLATFORMS

    - Linix (Mint)  -   tbc
    - Linux (Other) -   Untested
    - Raspbarian    -   tbc
    - Windows       -   Currently Unsupported
    
# PREREQUISITES

    * python 3
    * pip
    * git

# INSTALLATION

    cd ~
    git clone https://github.com/automation-itspeedway-net/miot.git
    miot/setup

# UPGRADE

    cd ~/miot
    git pull
    ./setup
    
# CONFIGURE MQTT SERVER

This step is only required if you are installing MIOT on a different server than your MQTT broker or if you are using authentication (Which you should be).

    miot set mqtt.host 192.168.1.1
    miot set mqtt.port 1883
    miot set mqtt.username MyUsername
    miot set mqtt.password MyPassword
    
# HOW TO USE IT

miot is a command-line utility that is used to control and configure your home automation modules.

You can view the help screen by typing any one of the following:
    miot --help
    miot -h
    miot /?

The next set of commands are used to control the modules you have selected to be part of your system. You can choose to install/uninstall, enable/disable or start/stop their services.

    miot install <module>
    miot remove <module>
    miot uninstall <module>
    miot enable <module>
    miot disable <module>
    miot start <module>
    miot stop <module>
    miot restart <module>
    miot status <module>

The following commands are only available after you have run 'miot setup':

Configuration of miot is performed from the command line (not by editing huge configuration files):

    miot set <key> <value>
    miot get <key>
    
After configuring MQTT, you can send low level MQTT messages:

    miot send [host] <topic> <message>

You can also use *miot* to snoop on MQTT messages:

    miot snoop [host] <topic>

# DEVELOPMENT LOG
DATE         VER    STATUS
15 MAR 2019  0.0    Initial Outline
07 JUL 2020  0.1    PRE-ALPHA
10 JUL 2020  0.1.1  PRE-ALPHA
12 JUL 2020  0.1.2  PRE-ALPHA

    


CHANGES.md
# CHANGE CONTROL

DATE         VERSION  STATUS
15 MAR 2019  0.0      Initial Outline
07 JUL 2020  0.1      PLANNING
15 JUL 2020  0.2      PRE-ALPHA

## PLANNING RELEASES

DATE         VERSION  STATUS
07 JUL 2020  0.1      PLANNING
    - Added target functions and commands
10 JUL 2020  0.1.1    PLANNING
    - Replaced "-" with "_" in module names
    - Seperate setup shell script
    - Bug fixes
12 JUL 2020  0.1.2    PLANNING
    - Bug fixes

## PRE-ALPHA RELEASES

DATE         VERSION  STATUS
15 JUL 2020  0.2.0    PRE-ALPHA
    - Introduced windows batch files (non functional)
    - Bug fixes
18 JUL 2020  0.2.1    PRE-ALPHA
    - Fixed bug calling repl
25 JUL 2020  0.2.2
    - Bug fixes
25 JUL 2020  0.2.3
    - Updated header display to align REPL version.


test/test-config.py
# Test the config module
# (c) Copyright Si Dunford, Jul 2020

from miot import config

print( config.tostring() )

config.wipe()

# TEST CONFIG
print( "RESULT="+config.get( "mqtt.host", "xyz" ) )
print( "RESULT="+config.get( "mqtt.default.host", "abc" ) )
print( "RESULT="+config.get( "rss.feed", "123" ) )

config.set( "mqtt.host", "192.168.1.228" )
config.set( "mqtt.default.host", "192.168.1.253" )
config.set( "rss.itspeedway", {'url':"https://itspeedway.net"} )
config.set( "rss.bbc", {'url':"https://bbc.co.uk"} )

config.save()

print( config.tostring() )




miot/miot.sh
#!/bin/bash

# MODULAR INTERNET OF THINGS
# (c) Copyright Si Dunford, Jun 2020
VERSION = "0.2.1"
# https://github.com/automation-itspeedway-net/miot

echo "MODULAR IoT"
echo "VER: $VERSION"

show_help()
{
    echo "miot --help|-h|/?"
    echo "miot setup"
    echo "miot install <module>"
    echo "miot remove|uninstall <module>"
    echo "miot enable <module>"
    echo "miot disable <module>"
    echo "miot start <module>"
    echo "miot stop <module>"
    echo "miot restart <module>"
    echo "miot status <module>"
}


go_systemctl()
{
    sudo systemctl $1 miot_$2
}

go_repl()
{
    if [ ! -e ~/miot/venv/bin/miotrepl ]; then
        echo "Unknown command: $1"
        echo "- REPL has not been installed."
        echo "- Run 'miot setup' to install"
    else    
        cd ~/miot
        source ./venv/bin/activate
        shift
        miotrepl $@
    fi
}

go_install()
{
    if [ ! -d ~/miot/ ]; then
        mkdir ~/miot
    fi
    if [ ! "$2" = "" ]; then
        echo "Installing $2"
        cd ~/miot
        if ! git clone https://github.com/automation-itspeedway-net/miot_$2.git
        then
            echo >&2 "Failed to install"
            exit 1
        fi
        chmod u+rwx miot_$2/miot_$2.service
        sudo cp miot_$2/miot_$2.service /etc/systemd/system/
    else
        echo "Module not specified"
    fi
}

go_remove()
{
    if [ -d ~/miot/miot_$2 ]; then
        sudo systemctl stop miot_$2
        sudo systemctl disable miot_$2
        sudo rm /etc/systemd/system/miot_$2.service
        rm -r "~/miot/miot_$2/"
    fi
}

go_setup()
{
    if [ -d ~/miot ]; then
        echo "- Folder exists"
    else
        echo "- Creating folders"
        mkdir ~/miot
    fi
    cd ~/miot
    if [ -e ~/miot/venv/bin/python ]; then
        echo "- Virtual environment exists"
    else
        echo "- Building virtual environment"
        python3 -m venv venv
    fi
    source ./venv/bin/activate
    echo "- Installing core components"
    python -m pip install --upgrade miot
    echo "- Done"
}

while [ "$1" != "" ]; do
    case $1 in
        --help | -h | /? )
            show_help
            exit
            ;;
        disable | enable | start | stop | restart | reload | status )    
            go_systemctl
            exit
            ;;
        install )
            go_install
            exit
            ;;
        remove | uninstall )
            go_remove
            exit
            ;;
        setup )
            go_setup
            exit
            ;;
        * )
            go_repl
            exit
            ;;
    esac
    shift
done


miot/config.py
# CONFIG MODULE
# Part of the MIOT package
#
# (C) Copyright Si Dunford, JUN 2020
#
# VERSION: 1.0

""" CHANGE CONTROL
1.0  05 JUL 2020  Creation. 
                  load(), save(), get(), set(), tostring(), wipe()
                  __default(), __getkey(), __setkey()
"""

import json

CONFIG_FILE = "/etc/miot.cfg"
__data = {}

def __default():
    global __data
    print( "* loading default values" )
    __data = {
        'mqtt':{
            'default':{'host':'127.0.0.1','port':1883}
            }
        }

# Walk the dictionary until we identify the requested key.
def __getkey( keys, config, alt ):
    if len(keys)==0: return alt
    #print( "* finding "+keys[0]+" in "+str(config) )
    # Check for key
    if keys[0] in config:
        #print( "found "+keys[0]+" in "+str(config) )
        #print( keys[1:] )
        if len(keys)==1: return config[keys[0]]
        return __getkey( keys[1:], config[keys[0]], alt )
    
    # Check for 'default'
    #print( "Checking for default value" )
    if 'default' in config:
        return __getkey( keys, config['default'], alt )
    
    #print( "failed to find "+keys[0]+" in "+str(config) )
    # Failed to find key
    return alt

def __setkey( keys, config, value ):
    if len(keys)==0: return

    print( "* finding "+keys[0]+" in "+str(config) )
    # Check for key
    if keys[0] in config:
        print( "found "+keys[0]+" in "+str(config) )
        print( keys[1:] )
        if len(keys)==1:
            print( "THINK THIS IS IT" )
            config[keys[0]]=value
            return
        __setkey( keys[1:], config[keys[0]], value )
        return

    # Check for 'default'
    print( "Checking for default value" )
    if 'default' in config:
        print( "Default found" )
        __setkey( keys, config['default'], value )
        return

    # Failed to find key
    print( "failed to find "+keys[0]+" in "+str(config) )
    if len(keys)==1:
        config[keys[0]]=value
        return
    config[keys[0]]={}
    __setkey( keys[1:], config[keys[0]], value )
    
def get( key, alt ):
    global __data
    print( "Searching for '"+key+"' ['"+alt+"']" )
    keys = key.split(".")
    print( keys )
    
    return __getkey( keys, __data, alt )
    
def set( key, value ):
    global __data
    print("")
    print( "Setting '"+key+"' to '"+str(value)+"'" )
    keys = key.split(".")
    print( keys )
    __setkey( keys, __data, value )
    print( "-RESULT")
    print( json.dumps( __data ) )

def load():
    global __data
    print( "LOADING" )
    try:
        with open( CONFIG_FILE, 'r') as cfg:
            __data = json.load(cfg)
    except:
        print( "Failed to load config" )
        #__default()
        pass
    if __data == {}:
        __default()
        
def save():
    global __data
    #__data={}
    print( "SAVING" )
    print( json.dumps( __data ) )
    with open( CONFIG_FILE, 'w' ) as cfg:
        json.dump( __data, cfg, indent=4, sort_keys=True )

def wipe():
    global __data
    __data={}
    save()
    
def tostring( indent=4, sort=True ):
    return json.dumps( __data, indent=indent, sort_keys=sort )
    
""" INITIALISE MODULE """

print( "config.init()" )
load()
print( json.dumps( __data ) )



miot/miot.py
# Modular Internet of Things, Core functionality
#
# FILE:     miot.py
# AUTHOR:   (c) Copyright Si.Dunford, 2019
# VERSION:  19.03.26.1751
# DATE:     26 March 2019
# STATE:    BETA
# LICENSE:  MIT License
#
# CHANGE LOG:
# 15 MAR 2019  0.0.1  Initial build
# 25 MAR 2019  0.1.1  Revised event handlers
# 26 MAR 2019  0.1.2  Bug fixes / Removed some debug output
#              0.1.3  Added config.getfloat(), config.set() and config.write()
# 27 MAR 2019  0.2.1  New version control, move from sandbox to miot
# 05 JUL 2020  0.3.1  Revision 3

# THIRD PARTY MODULES
import paho.mqtt.client as paho_mqtt
import configparser as configParser

# Default connect method
def mqtt_connect( client, userdata, flags, rc ):
    print "Connected"
    if rc==0:    # Connected successfully
        # Subscribing in on_connect() means that if we lose the connection and
        # reconnect then subscriptions will be renewed.
        print "Listening on", userdata.topic
        client.subscribe( userdata.topic )
    else:
        print("Connection refused: Result code "+str(rc))
        if rc==1:
            print "  Incorrect protocol version"
        elif rc==2:
            print "  Invalid client identifier"
        elif rc==3:
            print "  Server unavailable"
        elif rc==4:
            print "  Bad username or password"
        elif rc==5:
            print "  Not authorised"
        else:
            print "Result code",rc

# WRAPPER FOR CONFIGPARSER
class Config:
    def __init__( self, filename ):
        self.filename = filename
        self.config = configParser.ConfigParser()
        self.config.read( filename )

    def get( self, section, key, default='' ):
        if self.config.has_section( section ):
            return self.config[section].get( key, default )
        print '::CONFIG:: Section',section,'missing'
        return default

    def getint( self, section, key, default=0 ):
        if self.config.has_section( section ):
            return self.config[section].getint( key, default )
        print '::CONFIG:: Section',section,'missing'
        return default

    def getfloat( self, section, key, default=0.0 ):
        if self.config.has_section( section ):
            return self.config[section].getfloat( key, default )
        print '::CONFIG:: Section',section,'missing'
        return default

    def set( self, section, key, value ):
        if not self.config.has_section( section ):
            self.config.add_section( section )
        self.config.set( section, key, value )

    def write( self ):
        with open(self.filename, 'wb') as configfile:
            self.config.write(configfile)

# WRAPPER FOR PAHO MQTT
class MQTT:

    def __init__( self, config ):
        self.broker = config.get( 'MQTT','broker','127.0.0.1' )
        self.port = config.getint( 'MQTT','port',1883 )
        self.username = config.get( 'MQTT','username','' )
        self.password = config.get( 'MQTT','password','' )
        self.clientid = '' # Use random client ID
        self.topic = "miot"
        self.on_connect = mqtt_connect
        self.on_disconnect = None
        self.on_message = None
        self.on_publish = None
        self.on_subscribe = None

    def forever( self, topic='', message_handler=None ):
        # Process arguments
        if not topic=='':
            self.topic = topic
        if not message_handler==None:
            self.on_message=message_handler

        # Create MQTT object with optional authentication
        self.mqtt = paho_mqtt.Client( self.clientid )
        if not self.username=='':
            print 'Using Authentication...'
            self.mqtt.username_pw_set(username=self.username,password=self.password)

        # Set userdata to self
        self.mqtt.user_data_set(self)

        # Assign event handlers
        if not self.on_connect==None:
            self.mqtt.on_connect=self.on_connect
        if not self.on_disconnect==None:
            self.mqtt.on_disconnect=self.on_disconnect
        if not self.on_message==None:
            self.mqtt.on_message=self.on_message
        if not self.on_publish==None:
            self.mqtt.on_publish=self.on_publish
        if not self.on_subscribe==None:
            self.mqtt.on_subscribe=self.on_subscribe

        # Connect to MQTT broker
        self.mqtt.connect( self.broker, self.port )
        self.mqtt.loop_forever()


# CONFIG FILE
CONFIGFILE = 'config.ini'
#os.path.join( 'etc','opt','miot','config.ini' )
config = Config( CONFIGFILE )

# MQTT
mqtt = MQTT( config )


